activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -191,6 +191,19 @@ def pick(*keys)
   def compact_blank
     reject(&:blank?)
   end
+
+  # Returns a new +Array+ where the order has been set to that provided in the +series+, based on the +key+ of the
+  # objects in the original enumerable.
+  #
+  #   [ Person.find(5), Person.find(3), Person.find(1) ].in_order_of(:id, [ 1, 5, 3 ])
+  #   => [ Person.find(1), Person.find(5), Person.find(3) ]
+  #
+  # If the +series+ include keys that have no corresponding element in the Enumerable, these are ignored.
+  # If the Enumerable has additional elements that aren't named in the +series+, these are not included in the result.
+  def in_order_of(key, series)
+    indexed_by_key = index_by(&key)
+    series.filter_map { |s| indexed_by_key[s] }
```

Comment:
I think `Hash#values_at` would do as well:

```suggestion
    indexed_by_key.values_at(*series).compact
```

And it's a bit faster:

<details>
  <summary>Benchmark Script</summary>

  ```ruby
  require "benchmark/ips"
  require "active_support/core_ext/enumerable"

  VALUES = %w[ 1 12 123 1234 12345 123456 ]
  VALUES_BY_KEY = VALUES.index_by(&:length)
  SERIES = [ 1, 3, 5, 7, 2, 4, 6, 8 ]

  Benchmark.ips do |job|
    job.report("map + compact") do
      SERIES.map { |s| VALUES_BY_KEY[s] }.compact
    end

    job.report("values_at + compact") do
      VALUES_BY_KEY.values_at(*SERIES).compact
    end

    job.report("filter_map") do
      SERIES.filter_map { |s| VALUES_BY_KEY[s] }
    end

    job.compare!
  end
  ```
</details>

```
Comparison:
 values_at + compact:  1352561.4 i/s
       map + compact:   646014.7 i/s - 2.09x  (± 0.00) slower
          filter_map:   483476.8 i/s - 2.80x  (± 0.00) slower
```